Servers/controllers/users.controller.ts

```diff
@@ -0,0 +1,211 @@
+import { Request, Response } from "express";
+
+import { User } from "../models/user.model";
+import { UserOut } from "../dtos/userOut.dto";
+import { UserIn } from "../dtos/userIn.dto";
+
+import { writeFile } from "fs/promises";
+import bcrypt from "bcrypt";
+import { getUserOut, getUsers } from "../utils/users.utils";
+import { IUserAuthRequest } from "../types/IUserAuthRequest";
+import { generateToken } from "../utils/jwt.utils";
+
+type UserParams = Partial<Pick<User, "id" | "name" | "email">>;
+
+async function getAllUsers(req: IUserAuthRequest, res: Response): Promise<any> {
+    try {
+        let users = await getUsers()
+        let usersOut = getUserOut(users) as UserOut[]
+
+        return res.json({ data: usersOut })
+    } catch (error: any) {
+        return res.status(500).json({
+            data: "internal server error",
+            errorDetails: error.toString()
+        })
+    }
+}
+
+async function findUser(req: IUserAuthRequest, res: Response): Promise<any> {
+    const params = req.query as UserParams
+    const users = await getUsers()
+
+    const user = users.find(u => {
+
+        for (let p of Object.keys(params)) {
+            if (u[p as keyof UserParams] !== params[p as keyof UserParams]) {
+                return false
+            }
+        }
+        return true
+    })
+
+    if (user === undefined) {
+        return res.status(404).json({ data: `user with ${Object.keys(params).map(p => `${p}: ${params[p as keyof typeof params]}`).join(', ')} not found` })
+    } else {
+        return res.json({ data: getUserOut(user) as UserOut })
+    }
+}
+
+async function getUserFromId(req: IUserAuthRequest, res: Response): Promise<any> {
+    // assuming the userId will be integer parsable
+    const userId = parseInt(req.params.id);
+    try {
+        const users = await getUsers()
+        let found = false
+
+        for (let u of users) {
+            if (u.id === userId) {
+                return res.json({ data: getUserOut(u) as UserOut })
+            }
+        }
+
+        return res.status(404).json({ data: `user with id: ${userId} not found` })
+
+    } catch (error: any) {
+        return res.status(500).json({
+            data: "internal server error",
+            errorDetails: error.toString()
+        })
+    }
+}
+
+async function createUser(req: IUserAuthRequest, res: Response): Promise<any> {
+    const userIn = req.body as UserIn
+    const users = await getUsers()
+    const usersCtr = users.length
+
+    for (let u of users) {
+        if (u.email === userIn.email) {
+            return res.status(400).json({ data: `user with email: ${userIn.email} already exists` })
+        }
+    }
+
+    const user: User = {
+        id: usersCtr + 1,
+        created_at: Date.now(),
+        last_login: -1,
+        role_id: 1,
+        name: userIn.name,
+        email: userIn.email,
+        password: await bcrypt.hash(userIn.password, process.env.SALT as string)
```

Comment:
@HarshP4585 Thank you for considering the security point in your code 🙏🏻
Now, since this a very important application and we need to apply the security and safety things as strong as possible,
therefore I think we'd better make these suggestions